snapcraft/parts/plugins/kernel.py

```diff
@@ -0,0 +1,1683 @@
+        elif self._part_info.project_info.target_arch == "amd64":
+            self.deb_arch = "amd64"
+        else:
+            logger.error("Unknown deb architecture!!!")
+
+    def _check_cross_compilation(self) -> None:
+        if self._cross_building:
+            self.make_cmd.append(f"ARCH={self.kernel_arch}")
+            self.make_cmd.append("CROSS_COMPILE=${CRAFT_ARCH_TRIPLET}-")
+
+    def _set_kernel_targets(self) -> None:
+        if not self.options.kernel_image_target:
+            self.kernel_image_target = _default_kernel_image_target[self.deb_arch]
+        elif isinstance(self.options.kernel_image_target, str):
+            self.kernel_image_target = self.options.kernel_image_target
+        elif self.deb_arch in self.options.kernel_image_target:
+            self.kernel_image_target = self.options.kernel_image_target[self.deb_arch]
+
+        self.make_targets = [self.kernel_image_target, "modules"]
+        self.make_install_targets = [
+            "modules_install",
+            "INSTALL_MOD_STRIP=1",
+            "INSTALL_MOD_PATH=${CRAFT_PART_INSTALL}",
+        ]
+        if self.options.kernel_device_trees:
+            self.dtbs = [f"{i}.dtb" for i in self.options.kernel_device_trees]
+            if self.dtbs:
+                self.make_targets.extend(self.dtbs)
+        elif self.kernel_arch in ("arm", "arm64", "riscv64"):
+            self.make_targets.append("dtbs")
+            self.make_install_targets.extend(
+                ["dtbs_install", "INSTALL_DTBS_PATH=${CRAFT_PART_INSTALL}/dtbs"]
+            )
+        self.make_install_targets.extend(self._get_fw_install_targets())
+
+    def _get_fw_install_targets(self) -> List[str]:
+        if not self.options.kernel_with_firmware:
+            return []
+
+        return [
+            "firmware_install",
+            "INSTALL_FW_PATH=${CRAFT_PART_INSTALL}/lib/firmware",
+        ]
+
+    def _get_initrd_kernel_modules(self) -> List[str]:
+        # collect list of ko to install to the initrd
+        initrd_installed_kernel_modules = ""
+        initrd_configured_kernel_modules = ""
+        if self.options.kernel_initrd_modules:
+            initrd_installed_kernel_modules = (
+                f"{' '.join(self.options.kernel_initrd_modules)}"
+            )
+        if self.options.kernel_initrd_configured_modules:
+            initrd_configured_kernel_modules = (
+                f"{' '.join(self.options.kernel_initrd_configured_modules)}"
+            )
+        return [
+            "# listf of kernel modules to be installed in initrd",
+            f'initrd_installed_kernel_modules="{initrd_installed_kernel_modules}"',
+            "# listf of kernel modules in initrd to be auto loaded by",
+            "# any module in this list implies to be added to initrd",
+            f'initrd_configured_kernel_modules="{initrd_configured_kernel_modules}"',
+        ]
+
+    def _link_files_fnc_cmd(self) -> List[str]:
+        return [
```

Comment:
Maybe dedenting could make this easier to read, but that can be done in a future refactoring PR.

Reply:
If you are referring to lines like `"\t\tfi"`
Then those are there on purpose. It makes generated `build.sh` a lot more readable.
Considering plugins does many things, there is an effort to make generated `build.sh` readable, in case there's a build error which needs to be debugged.
I have changed the format a bit, can you please check if this makes it more readable.

Reply:
The formatting should be fine, we can improve this later if need arises.